File: 1203_Sort_Items_by_Groups_Respecting_Dependencies/solution.cpp
/*
1203. Sort Items by Groups Respecting Dependencies

Level: Hard

https://leetcode.com/problems/sort-items-by-groups-respecting-dependencies
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <set>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 

  所有无人认领的小组分配一个组号，这个组内只有这一个项目。
  接下来统计一下每个组之间边关系和入度，因为是统计组之间的，所有组内依赖暂时不考虑，还需要考虑的是如果两个组之间有多条边，我们只计算一次。
  同时，统计每一个组内部的边关系和入度，因为是统计组内的，所以组间依赖关系暂时不考虑。
  对组进行拓扑排序，看一下是否有冲突，如果有冲突就返回，否则继续。
  按照组的拓扑序，对每一个组内部进行拓扑排序。
*/

class Solution
{
public:
  vector<int> sortItems(int n, int m, vector<int> &group, vector<vector<int>> &beforeItems)
  {
    unordered_map<int, set<int>> group_to_item_hash;
    // item items do not belong to group to new group
    for (int i = 0; i < n; i++)
    {
      if (group[i] == -1)
      {
        group[i] = m++;
      }
      group_to_item_hash[group[i]].insert(i);
    }

    // build graph
    vector<set<int>> group_graph(m, set<int>()), item_graph(n, set<int>());
    vector<int> group_indegree(m, 0);
    vector<int> item_indegree(n, 0);

    for (int i = 0; i < n; i++)
    {
      if (beforeItems[i].size() == 0)
        continue;
      for (auto from : beforeItems[i]) // i is to
      {
        if (group[from] != group[i]) // beforeItem is in different group
        {
          if (group_graph[group[from]].count(group[i]) == 0)
          {
            group_graph[group[from]].insert(group[i]);
            group_indegree[group[i]]++;
          }
        }
        else
        {
          item_graph[from].insert(i);
          item_indegree[i]++;
        }
      }
    }

    vector<int> group_ordered = tpSort(group_graph, group_indegree);

    if (group_ordered.size() != m)
      return {};

    vector<int> ans;
    for (auto group_id : group_ordered)
    {
      queue<int> q;
      for (auto item_id : group_to_item_hash[group_id])
      {
        if (item_indegree[item_id] == 0)
          q.push(item_id);
      }
      int count = 0;
      while (!q.empty())
      {
        int p = q.front();
        q.pop();
        ans.push_back(p);
        count++;
        for (auto next : item_graph[p])
        {
          item_indegree[next]--;
          if (item_indegree[next] == 0)
            q.push(next);
        }
      }
      if (count != group_to_item_hash[group_id].size())
        return {};
    }
    return ans;
  }

  vector<int> tpSort(vector<set<int>> &graph, vector<int> &in_degree)
  {
    vector<int> sorted;
    queue<int> q;

    for (int i = 0; i < in_degree.size(); i++)
    {
      if (in_degree[i] == 0)
      {
        q.push(i);
      }
    }

    while (!q.empty())
    {
      int cur = q.front();
      q.pop();
      sorted.push_back(cur);

      for (int next : graph[cur])
      {
        in_degree[next]--;
        if (in_degree[next] == 0)
        {
          q.push(next);
        }
      }
    }
    return sorted;
  }
};

File: 1203_Sort_Items_by_Groups_Respecting_Dependencies/solution.java
/*
1203. Sort Items by Groups Respecting Dependencies

Level: Hard

https://leetcode.com/problems/sort-items-by-groups-respecting-dependencies
*/
package cn.huang.leetcode;

import java.util.HashMap;
import java.util.Map;
/*
  Solution: 

  https://www.youtube.com/watch?v=nE97bUETsCM

  1. topoligical sort items.
  2. topolobical srot group
  3. put sorted items to its group with the output of step 1 and 2
  4. iterate gropu in order and add items in each boutp to the result array
*/
class Solution {
  public int[] sortItems(int n, int m, int[] group, List<List<Integer>> beforeItems) {
    HashMap<Integer, List<Integer>> gGraph = new HashMap<>();
    HashMap<Integer, List<Integer>> iGraph = new HashMap<>();

    // item items do not belong to group to new group
    for(int i=0; i< group.length; i++){
      if(group[i] == -1)
      {
        // i-th item belongs to no group
        group[i] = m++; 
      }
    }

    int[] itemInDegree = new int[n];
    int[] groupInDegree = new int[m];     // m is increased abrove.

    // build item graph
    for(int  to =0; to < beforeItems.size(); to ++)
    {
      int toGroup = group[to];

      for(int from: beforeItems.get(to)){
        itemInDegree[to]++;
        if(!iGraph.containsKey(from))
        {
          iGraph.put(from, new ArrayList<Integer>());
        }
        iGraph.get(from).add(to);
      }
    }
    // build group graph
    for(int  to =0; to < group.length; to++)
    {
      int toGroup = group[to];

      for(int from: beforeItems.get(to)){
        int fromGroup = group[to];
        if(!gGraph.containsKey(from))
        {
          gGraph.put(fromGroup, new ArrayList<Integer>());
        }
       if(fromGroup != toGroup){
          groupInDegree[toGroup]++;
          gGraph.get(fromGroup).add(to);
        }
      }
    }

    List<Integer> iList = tpSort(iGraph, itemInDegree, n);
        List<Integer> gList = tpSort(gGraph, groupInDegree, n);


    if(iList.size() == 0 || gList.size() ==0)
      return new int[0];


    HashMap<Integer, List<Integer>> groupedList = new HashMap<>();
    
    for(int item : iList){
      int grp = group[item];
      if(!groupedList.containsKey(grp))
      {
        groupedList.put(grp, new ArrayList<Integer>());
      }
      groupedList.get(grp).add(item);
    }

    int i =0 ;
    int[] ans = new int[0];
    for(int grp: gList){
      if(groupedList.containsKey(grp))
      {
        for(int item: groupedList.get(grp)){
          ans[i] = item;
          i++;
        }
      }
    }
    return ans;
  }

  public List<Integer> tpSort(HashMap<Integer, List<Integer>> graph, int[] inDegree, int count){
    List<Integer> ans = new ArrayList<>();

    Queue<Integer> q=new LinkedList<>();

    for(int i =0; i < inDegree.length; i++){
      if (inDegree[i] == 0){
        q.offer(i);
      }
    }
    while(!q.isEmpty()){
      int cur = q.poll();
      ans.add(cur);
      if(graph.containsKey(cur)){
        for(int next: graph.get(cur)){
          inDegree[next]--;
          if(inDegree[next] ==0){
            q.offer(next);
          }
        }
      }
    }

    return count == ans.size() ? ans : new ArrayList<>();
  }
} 